File: solutions/nim/dnd-character/1/dnd_character.nim
import std/random
import std/sequtils
import std/algorithm
import std/math

type Character* = object
  strength*: int
  dexterity*: int
  constitution*: int
  intelligence*: int
  wisdom*: int
  charisma*: int
  hitpoints*: int

const throwsPerAbility = 4
const summedThrows = 3

proc ability*(): int =
  return (1 .. throwsPerAbility).mapIt(rand(5) + 1).sorted()[1 .. summedThrows].sum()

proc modifier*(n: int): int =
  return floorDiv(n - 10, 2)

proc initCharacter*(): Character =
  let constitution = ability()
  Character(
    strength: ability(),
    dexterity: ability(),
    constitution: constitution,
    intelligence: ability(),
    wisdom: ability(),
    charisma: ability(),
    hitpoints: 10 + modifier(constitution),
  )

File: solutions/lua/dnd-character/2/dnd-character.lua
local function ability(scores)
  local min = 6
  local sum = 0
  for _, value in ipairs(scores) do
    if value < min then
      min = value
    end
    sum = sum + value
  end
  return sum - min
end

local function roll_dice()
  local rolls = {}
  for i = 1, 4 do
    rolls[i] = math.random(6)
  end
  return rolls
end

local function modifier(input)
  return math.floor((input - 10) / 2)
end

local Character = {}
function Character:new(name)
  local strength = ability(roll_dice())
  local dexterity = ability(roll_dice())
  local constitution = ability(roll_dice())
  local intelligence = ability(roll_dice())
  local wisdom = ability(roll_dice())
  local charisma = ability(roll_dice())
  local hitpoints = 10 + modifier(constitution)
  return {
    name = name,
    strength = strength,
    dexterity = dexterity,
    constitution = constitution,
    intelligence = intelligence,
    wisdom = wisdom,
    charisma = charisma,
    hitpoints = hitpoints
  }
end

return {
  Character = Character,
  ability = ability,
  roll_dice = roll_dice,
  modifier = modifier
}
